design/specs/visual-query-builder.md

```diff
@@ -0,0 +1,236 @@
+
+---
+Wireframes
+
+[Adding Columns](https://share.balsamiq.com/c/oyxTXxqSh8rLqU3JY71DWY.png)
+
+[Added Column](https://share.balsamiq.com/c/7U9yahZtYWX2G5obkyivoz.png)
+
+### 2.2 Input Column Sources
+
+Once the input columns have been added, the user can inspect them by selecting each from the result table or the list of columns in the query configuration panel, respectively. A section titled 'Source' will be included in the details for each column. The source column, table, and links will all be listed in this portion of the panel.
+
+The source will include `Links` only for data sourced via tables connected to or from the base table.
+
+---
+Wireframes
+
+[Column Sources](https://share.balsamiq.com/c/ewCxvkjS8VoScKfgb9hFos.png)
+
+### 2.3 Adding Formulas
+
+Formulas are used to generate new columns based on different parameters. To access the list of formulas, the user start the `Add Column` process and selects the option `From Formula` at the top of the inspector panel. Selecting a formula will open a form that users can fill out to determine the values of the new column.
+
+Depending on the selected formula, different settings will be available.
+More on formulas and specific details for each will be covered on a separate issue.
+
+Columns generated from formulas will display a formula icon indicator in the column header.
+
+---
+Wireframes
+
+[Added Formula](https://share.balsamiq.com/c/vWfJ9sUYWJxed5Zo5WtPGi.png)
+
+### 2.4 Filtering Input Column Values
+
+Columns that link to multiple records can have filters applied to them to retrieve only values that match user-specified criteria. Multiple filters are allowed for each input column.
+
+Filter options will be determined by the data type of the input column.
+
+Filters can be added by clicking on the `Add Filter` option from the query configuration panel or directly from the results table column header menu.
+
+Columns with filters applied will display a filter icon indicator in the column header.
+
+---
+Wireframes
+
+[Added Input Filter](https://share.balsamiq.com/c/tK2hZy6FB4zVYpN56ejpBk.png)
+
+### 2.5 Aggregating Input Column Values
+
+In cases where a link may refer to multiple values, input columns can be aggregated. In such cases, the system will perform automatic aggregation based on the data type of the referenced column. Users can change the type of aggregation at any time.
+
+Aggregated columns will have an aggregation icon indicator in the column header.
+
+---
+Wireframes
+
+[Added Input Aggregation](https://share.balsamiq.com/c/fNcbvQ1q5xMyni5JxBg3Rc.png)
+
+### 2.6 Applying a Formula to an Input Column
+
+Direct input columns can also be transformed into formulas. The available formulas will depend on the input column type.
+
+To transform the input column into a formula, the user will select the `Apply Formula` option in the column header menu.
+
+---
+Wireframes
+
+[Applying a Formula](https://share.balsamiq.com/c/nEzYaNKNTSv9EFwzwtfSof.png)
```

Comment:
I think it's best to have a single workflow:

* Either user adds a formula and chooses the columns (My choice), or
* User adds a column and chooses the formula..

A formula can modify the column entirely, and already applied filters and aggregations would no longer be applicable. A user adding a formula to an existing column might not understand that the filters & aggregations would get removed.

Overall, I think this additional flexibility would be troublesome for the user and for us to maintain in the long run.

Reply:
@ghislaineguerin As we discussed yesterday, I think we should remove formulas from this spec entirely and have a separate spec for integrating formulas once we finish the design for https://github.com/centerofci/mathesar/issues/1252.